File: sha1.go
package main

import (
	"crypto/sha1"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

// generateCombinations generates all possible combinations of characters
func generateCombinations(charset string, length int, prefix string, target string, knownPrefix string) (string, int64) {
	var count int64 = 0
	startTime := time.Now()

	// Create a function to recursively generate combinations
	var generate func(current string, depth int) string
	generate = func(current string, depth int) string {
		// Base case: if we've reached the desired length
		if depth == length {
			count++

			// Generate the potential flag
			potentialFlag := knownPrefix + current

			// Calculate SHA1 hash
			hash := sha1.Sum([]byte(potentialFlag))
			hashStr := fmt.Sprintf("%x", hash)

			// Check if hash matches target
			if hashStr == target {
				return potentialFlag
			}

			// Print progress every million attempts
			if count%1000000 == 0 {
				now := time.Now()
				elapsed := now.Sub(startTime).Seconds()
				rate := float64(count) / elapsed
				fmt.Printf("Tried %d combinations... (%.2f attempts/second)\n", count, rate)
			}

			return ""
		}

		// Try each character in the charset
		for i := 0; i < len(charset); i++ {
			result := generate(current+string(charset[i]), depth+1)
			if result != "" {
				return result
			}
		}

		return ""
	}

	// Start generation with empty string
	result := generate("", 0)

	return result, count
}

func main() {
	fmt.Println("SHA1 Hash Brute Force Tool")
	fmt.Println("==========================")

	// Hash options
	fmt.Println("\nSelect the hash type:")
	fmt.Println("1. SHA1")
	fmt.Println("2. SHA2")

	// Get target hash
	fmt.Print("Enter the SHA1 hash: ")
	var targetHash string
	fmt.Scanln(&targetHash)
	targetHash = strings.ToLower(targetHash)

	// Get known prefix
	fmt.Print("Enter the known prefix: ")
	var knownPrefix string
	fmt.Scanln(&knownPrefix)

	// Character set options
	fmt.Println("\nSelect character set for brute force:")
	fmt.Println("1. Numbers only (0-9)")
	fmt.Println("2. Lowercase letters only (a-z)")
	fmt.Println("3. Uppercase letters only (A-Z)")
	fmt.Println("4. Numbers and lowercase letters")
	fmt.Println("5. Numbers and uppercase letters")
	fmt.Println("6. All letters (a-z, A-Z)")
	fmt.Println("7. All alphanumeric characters (0-9, a-z, A-Z)")
	fmt.Println("8. Numbers and lowercase letters only (0-9, a-z)")

	// Get character set choice
	fmt.Print("Enter your choice (1-8): ")
	var choice string
	fmt.Scanln(&choice)

	// Define character sets
	digits := "0123456789"
	lowercaseLetters := "abcdefghijklmnopqrstuvwxyz"
	uppercaseLetters := "ABCDEFGHIJKLMNOPQRSTUVWXYZ"

	var charset string

	// Set charset based on user choice
	switch choice {
	case "1":
		charset = digits
	case "2":
		charset = lowercaseLetters
	case "3":
		charset = uppercaseLetters
	case "4":
		charset = digits + lowercaseLetters
	case "5":
		charset = digits + uppercaseLetters
	case "6":
		charset = lowercaseLetters + uppercaseLetters
	case "7":
		charset = digits + lowercaseLetters + uppercaseLetters
	case "8":
		charset = digits + lowercaseLetters
	default:
		fmt.Println("Invalid choice. Using all alphanumeric characters.")
		charset = digits + lowercaseLetters + uppercaseLetters
	}

	// Get maximum length
	fmt.Print("Enter maximum length of characters to try (default: 8): ")
	var maxLengthStr string
	fmt.Scanln(&maxLengthStr)

	maxLength := 8
	if maxLengthStr != "" {
		var err error
		maxLength, err = strconv.Atoi(maxLengthStr)
		if err != nil {
			fmt.Println("Invalid input. Using default value of 8 characters.")
			maxLength = 8
		}
	}

	fmt.Printf("Starting brute force with prefix: %s\n", knownPrefix)
	fmt.Printf("Character set size: %d (%s)\n", len(charset), charset)
	fmt.Printf("Trying combinations up to %d characters\n", maxLength)

	startTime := time.Now()
	var totalAttempts int64 = 0

	// Try different lengths
	for length := 1; length <= maxLength; length++ {
		fmt.Printf("Trying %d character combinations...\n", length)

		result, attempts := generateCombinations(charset, length, "", targetHash, knownPrefix)
		totalAttempts += attempts

		if result != "" {
			elapsed := time.Since(startTime).Seconds()
			fmt.Printf("Flag found: %s\n", result)
			fmt.Printf("Completed in %.2f seconds after %d attempts\n", elapsed, totalAttempts)
			os.Exit(0)
		}
	}

	elapsed := time.Since(startTime).Seconds()
	fmt.Printf("Search completed in %.2f seconds after %d attempts\n", elapsed, totalAttempts)
	fmt.Println("Flag not found. Try increasing max_length or check your inputs.")
}

File: sha1.py
import hashlib
import itertools
import string
import time

def brute_force_sha1(target_hash, known_prefix, char_set, max_length=8):
    """
    Brute force a SHA1 hash by trying different alphanumeric combinations
    """
    target_hash = target_hash.lower()
    start_time = time.time()
    attempts = 0

    print(f"Starting brute force with prefix: {known_prefix}")
    print(f"Character set size: {len(char_set)} ({char_set})")
    print(f"Trying combinations up to {max_length} characters")

    # Try combinations of increasing length
    for length in range(1, max_length + 1):
        print(f"Trying {length} character combinations...")

        # Generate all possible combinations of the current length
        for combo in itertools.product(char_set, repeat=length):
            suffix = ''.join(combo)
            potential_flag = f"{known_prefix}{suffix}"

            # Calculate SHA1 hash of the potential flag
            sha1_hash = hashlib.sha1(potential_flag.encode()).hexdigest()

            attempts += 1

            # Check if the hash matches the target
            if sha1_hash == target_hash:
                elapsed = time.time() - start_time
                print(f"Completed in {elapsed:.2f} seconds after {attempts:,} attempts")
                return potential_flag

            # Print progress every million attempts
            if attempts > 0 and attempts % 1000000 == 0:
                elapsed = time.time() - start_time
                rate = attempts / elapsed if elapsed > 0 else 0
                print(f"Tried {attempts:,} combinations... ({rate:.2f} attempts/second)")

    elapsed = time.time() - start_time
    print(f"Search completed in {elapsed:.2f} seconds after {attempts:,} attempts")
    return None

def main():
    # Get inputs from user
    target_hash = input("Enter the SHA1 hash: ")
    known_prefix = input("Enter the known prefix: ")

    # Character set options
    print("\nSelect character set for brute force:")
    print("1. Numbers only (0-9)")
    print("2. Lowercase letters only (a-z)")
    print("3. Uppercase letters only (A-Z)")
    print("4. Numbers and lowercase letters")
    print("5. Numbers and uppercase letters")
    print("6. All letters (a-z, A-Z)")
    print("7. All alphanumeric characters (0-9, a-z, A-Z)")

    choice = input("Enter your choice (1-7): ")

    # Define character set based on user choice
    if choice == '1':
        char_set = string.digits
    elif choice == '2':
        char_set = string.ascii_lowercase
    elif choice == '3':
        char_set = string.ascii_uppercase
    elif choice == '4':
        char_set = string.digits + string.ascii_lowercase
    elif choice == '5':
        char_set = string.digits + string.ascii_uppercase
    elif choice == '6':
        char_set = string.ascii_letters
    elif choice == '7':
        char_set = string.digits + string.ascii_letters
    else:
        print("Invalid choice. Using all alphanumeric characters.")
        char_set = string.digits + string.ascii_letters

    try:
        max_length = int(input("Enter maximum length of characters to try (default: 8): ") or "8")
    except ValueError:
        print("Invalid input. Using default value of 8 characters.")
        max_length = 8

    # Try to find the flag
    flag = brute_force_sha1(target_hash, known_prefix, char_set, max_length)

    if flag:
        print(f"Flag found: {flag}")
    else:
        print("Flag not found. Try increasing max_length or check your inputs.")

if __name__ == "__main__":
    main()

